src/Mongolid/DataMapper/DataMapper.php

```diff
@@ -0,0 +1,280 @@
+<?php
+namespace Mongolid\DataMapper;
+
+use Mongolid\Schema;
+use Mongolid\Container\Ioc;
+use Mongolid\Connection\Pool;
+use Mongolid\Cursor\Cursor;
+use MongoDB\Collection;
+use MongoDB\BSON\ObjectID;
+
+/**
+ * The DataMapper class will abstract how an Entity is persisted and retrieved
+ * from the database.
+ * The DataMapper will always use a Schema trought the SchemaMapper to parse the
+ * document in and out of the database.
+ *
+ * @package  Mongolid
+ */
+class DataMapper
+{
+    /**
+     * Name of the schema class to be used
+     * @var string
+     */
+    public $schemaClass = Schema::class;
+
+    /**
+     * Schema object. Will be set after the $schemaClass
+     * @var Schema
+     */
+    public $schema;
+
+    /**
+     * Connections that are going to be used to interact with the database
+     * @var Pool
+     */
+    protected $connPool;
+
+    /**
+     * @param Pool        $connPool The connections that are going to be used to interact with the database
+     * @param Schema|null $schema   The Schema class of the model that is going go be persisted
+     */
+    public function __construct(Pool $connPool)
+    {
+        $this->connPool = $connPool;
+    }
+
+    /**
+     * Upserts the given object into database. Returns success if write concern
+     * is acknowledged.
+     *
+     * @return boolean Success (but always false if write concern is Unacknowledged)
+     */
+    public function save($object)
+    {
+        $data = $this->parseToDocument($object);
+
+        $result = $this->getCollection()->updateOne(
+            ['_id' => $data['_id']],
+            ['$set' => $data],
+            ['upsert' => true]
+        );
+
+        return (bool) ($result->getModifiedCount() + $result->getUpsertedCount());
+    }
+
+    /**
+     * Inserts the given object into database. Returns success if write concern
+     * is acknowledged. Since it's an insert, it will fail if the _id already
+     * exists.
+     *
+     * @return boolean Success (but always false if write concern is Unacknowledged)
+     */
+    public function insert($object): bool
+    {
+        if (! $object->_id) {
```

Comment:
Can you move this statement to method `parseToDocument`, because you repeat it at all CRUD operations.